### ISSUE TITLE
Training reproduction gives lower performance than released checkpoint

### ISSUE THREAD

**Issue by wdseh** [JinLi998/CoSMo3D#13]
## Training reproduction gives lower performance than released model

Hi, thanks for releasing the training code and dataset!

We successfully completed a 200-epoch training run using the released pipeline. The training runs normally, but the reproduced model performs noticeably worse than the reported/released model, especially on **3DCoMPaT coarse**.

### Performance comparison

Using exactly the same evaluation pipeline:

```text
3DCoMPaT coarse
  Paper / released model : 47.18
  Our reproduction       : 34.19
  Gap                    : -12.99

ShapeNetPart
  Paper / released model : 36.11
  Our reproduction       : 32.63
  Gap                    : -3.48
```

As a sanity check, the released `ours_final.pth` reproduces the expected **47.18 / 36.11** scores using the same evaluation pipeline, so the performance gap appears to come from training rather than evaluation.

The released `ours_final.pth` reports `epoch=180`. We therefore also evaluated our checkpoints from epoch 150 to 200:

```text
Epoch    3DCoMPaT coarse mIoU
150      33.77
160      33.82
170      34.71   <- best
175      34.59
180      33.66
185      34.58
190      34.32
195      34.36
200      34.19
```

Our best checkpoint in this range is **34.71 at epoch 170**, while epoch 180 gives **33.66**. Therefore, the gap does not seem to be explained simply by checkpoint selection.

### Training setup

We followed the released configuration as closely as possible:

* 4 × RTX 3090 24GB
* PyTorch 2.5.1 + CUDA 12.1
* FlashAttention enabled
* `n_epoch=200`
* `lr=5e-4`
* `eta_min=5e-5`
* `drop_canoncolor_last_n_epochs=30`
* pretrained from `orgfind3d.pth`
* released `trainingdata`

The main configuration difference is the **batch size**.

The released command uses `batch_size=32` per GPU, but this caused OOM on our 24GB RTX 3090s. We therefore used:

```text
batch_size      = 24 per GPU
global batch    = 96
number of GPUs  = 4
epochs          = 200
learning rate   = 5e-4
```

We kept the learning rate and number of epochs unchanged.

We understand that reducing the global batch size from 128 to 96 increases the number of optimizer updates per epoch even though each epoch still sees the full dataset, so this may affect the optimization trajectory.

### Small changes we made

For transparency, we made the following changes:

1. Added a `tqdm` progress bar for visualization only.

2. Loaded the same `google/siglip-base-patch16-224` model from a local directory because our server cannot reliably access Hugging Face.

3. Fixed a DDP issue in the canonical-color loss.

The released code contains:

```python
batch_size = len(pt_offset) - 1
```

However, `pt_offset` appears to contain one cumulative end offset per sample. With batch size 1, this makes the effective batch size 0 and causes the canonical-color branch to be skipped, resulting in unused `decoder.color_head` parameters under DDP.

We changed it to:

```python
batch_size = len(pt_offset)
```

and used a graph-connected zero when no valid canonical supervision exists.

After this change, both a 2-GPU DDP test and the full 4-GPU 200-epoch training completed successfully with `find_unused_parameters=False`.

4. We made small compatibility fixes to the released benchmark interface (`PointSemSegWithDecoder.forward(data)` and local SigLIP loading). These changes are applied equally when evaluating the released and reproduced checkpoints.

### Questions

Could you please help clarify:

1. Was the released `ours_final.pth` trained directly using the currently released `train_code/train_release.py` pipeline?

2. Is `batch_size=32` **per GPU** important for reproduction? Since we used batch size 24 while keeping 200 epochs and `lr=5e-4`, would you recommend adjusting the learning rate or using gradient accumulation to match the original effective batch size?

3. Is the following line in `loss_canonical_color.py` intentional?

```python
batch_size = len(pt_offset) - 1
```

Or should all entries in `pt_offset` be processed?

4. Were there any additional differences in the training setup used for the reported/released model, such as checkpoint selection, freezing schedule, loss implementation, or other hyperparameters?

Thanks again for your help!

**Comment by JinLi998** [comment]
> ## Training reproduction gives lower performance than released model
> Hi, thanks for releasing the training code and dataset!
> 
> We successfully completed a 200-epoch training run using the released pipeline. The training runs normally, but the reproduced model performs noticeably worse than the reported/released model, especially on **3DCoMPaT coarse**.
> 
> ### Performance comparison
> Using exactly the same evaluation pipeline:
> 
> ```
> 3DCoMPaT coarse
>   Paper / released model : 47.18
>   Our reproduction       : 34.19
>   Gap                    : -12.99
> 
> ShapeNetPart
>   Paper / released model : 36.11
>   Our reproduction       : 32.63
>   Gap                    : -3.48
> ```
> 
> As a sanity check, the released `ours_final.pth` reproduces the expected **47.18 / 36.11** scores using the same evaluation pipeline, so the performance gap appears to come from training rather than evaluation.
> 
> The released `ours_final.pth` reports `epoch=180`. We therefore also evaluated our checkpoints from epoch 150 to 200:
> 
> ```
> Epoch    3DCoMPaT coarse mIoU
> 150      33.77
> 160      33.82
> 170      34.71   <- best
> 175      34.59
> 180      33.66
> 185      34.58
> 190      34.32
> 195      34.36
> 200      34.19
> ```
> 
> Our best checkpoint in this range is **34.71 at epoch 170**, while epoch 180 gives **33.66**. Therefore, the gap does not seem to be explained simply by checkpoint selection.
> 
> ### Training setup
> We followed the released configuration as closely as possible:
> 
> * 4 × RTX 3090 24GB
> * PyTorch 2.5.1 + CUDA 12.1
> * FlashAttention enabled
> * `n_epoch=200`
> * `lr=5e-4`
> * `eta_min=5e-5`
> * `drop_canoncolor_last_n_epochs=30`
> * pretrained from `orgfind3d.pth`
> * released `trainingdata`
> 
> The main configuration difference is the **batch size**.
> 
> The released command uses `batch_size=32` per GPU, but this caused OOM on our 24GB RTX 3090s. We therefore used:
> 
> ```
> batch_size      = 24 per GPU
> global batch    = 96
> number of GPUs  = 4
> epochs          = 200
> learning rate   = 5e-4
> ```
> 
> We kept the learning rate and number of epochs unchanged.
> 
> We understand that reducing the global batch size from 128 to 96 increases the number of optimizer updates per epoch even though each epoch still sees the full dataset, so this may affect the optimization trajectory.
> 
> ### Small changes we made
> For transparency, we made the following changes:
> 
> 1. Added a `tqdm` progress bar for visualization only.
> 2. Loaded the same `google/siglip-base-patch16-224` model from a local directory because our server cannot reliably access Hugging Face.
> 3. Fixed a DDP issue in the canonical-color loss.
> 
> The released code contains:
> 
> batch_size = len(pt_offset) - 1
> However, `pt_offset` appears to contain one cumulative end offset per sample. With batch size 1, this makes the effective batch size 0 and causes the canonical-color branch to be skipped, resulting in unused `decoder.color_head` parameters under DDP.
> 
> We changed it to:
> 
> batch_size = len(pt_offset)
> and used a graph-connected zero when no valid canonical supervision exists.
> 
> After this change, both a 2-GPU DDP test and the full 4-GPU 200-epoch training completed successfully with `find_unused_parameters=False`.
> 
> 4. We made small compatibility fixes to the released benchmark interface (`PointSemSegWithDecoder.forward(data)` and local SigLIP loading). These changes are applied equally when evaluating the released and reproduced checkpoints.
> 
> ### Questions
> Could you please help clarify:
> 
> 1. Was the released `ours_final.pth` trained directly using the currently released `train_code/train_release.py` pipeline?
> 2. Is `batch_size=32` **per GPU** important for reproduction? Since we used batch size 24 while keeping 200 epochs and `lr=5e-4`, would you recommend adjusting the learning rate or using gradient accumulation to match the original effective batch size?
> 3. Is the following line in `loss_canonical_color.py` intentional?
> 
> batch_size = len(pt_offset) - 1
> Or should all entries in `pt_offset` be processed?
> 
> 4. Were there any additional differences in the training setup used for the reported/released model, such as checkpoint selection, freezing schedule, loss implementation, or other hyperparameters?
> 
> Thanks again for your help!

Thank you for your thorough ablation studies and timely feedback!

In fact, after releasing the code and dataset, we have run training with our released pipeline on A40 GPUs, and the obtained performance is even slightly higher than the numbers reported in our paper.

I will carefully go through all the detailed questions you raised and get back to you shortly.

**Comment by wdseh** [comment]
Thanks again for your reply. I rebuilt the reproduction environment and dataset from scratch, and also reviewed the existing closed issues before following up. I found a few concrete details that I would like to confirm.

**1. Exact training manifest (train.txt)**

The currently released trainingdata.tar.gz does not contain train.txt, while train_release.py expects --data_root/train.txt.

I reconstructed a manifest by keeping every sample directory containing the five files required by the released loader:

mask_labels.txt
mask2points.pt
points.pt
normals.pt
rgb.pt

This gives exactly 28,159 training samples:

13,809 coarse
14,350 fine

out of 28,700 directories in the archive, leaving 541 incomplete directories unused.

Could you confirm whether these 28,159 samples are exactly the training set used for ours_final.pth? If possible, could you also provide the original **train.txt** (and val.txt, if applicable), since the exact manifest/order is currently not included in the released archive?

**2. Loading orgfind3d.pth**

With a fresh download of the released orgfind3d.pth, I compared its state dict with the released CoSMo3D model:

orgfind3d keys            = 495
direct same-name matches  = 1
matches after "backbone." = 495

For example:

extractor.embedding.stem.conv.weight

in orgfind3d.pth corresponds to:

backbone.extractor.embedding.stem.conv.weight

in the full CoSMo3D model.

However, the current train_release.py loads it using:

model.load_state_dict(state, strict=False)

Could you confirm whether ours_final.pth was trained with this exact loading behavior? Or did the actual training run load the checkpoint into model.backbone / add the **backbone. prefix** before loading?

**3. Effective batch size of the released final model**

I also noticed your answer in issue #10 that the final model was trained on 8 GPUs with about 85.9 GB memory each.

Since the released code defines batch_size=32 per GPU, could you confirm whether the final ours_final.pth used:

8 GPUs × 32 samples/GPU = effective global batch size 256

with learning rate 5e-4?

This may be important for our reproduction, because on RTX 3090 GPUs, batch size 32 per GPU OOMs, so our previous run used 4 GPUs × 24 = global batch 96.

These checks were all performed again on a completely fresh clone, fresh checkpoints, fresh training archive, and a newly rebuilt environment, so they should not be caused by modifications from our previous reproduction setup.

Thank you again for taking the time to look into this.

**Comment by JinLi998** [comment]
Thank you very much for your thorough investigation and detailed feedback. Below are our answers to your questions:

1. **Training manifest & incomplete samples**
The released `ours_final.pth` originates from an internal pre‑release checkpoint. The public codebase has been re‑trained, yielding slightly better performance than the numbers reported in the paper.

There are indeed **541 corrupted samples** caused by upload failures in the published dataset archive. These samples were excluded during our retraining, which improves coarse‑set accuracy by about 1‑2 points compared with the original paper. We will re‑process and supplement the complete dataset and official `train.txt`.
`train.txt` simply stores sample paths such as `cosmo3d/dataset_/forfind3dtrain/coarse_b'00_020'`.

Also, we confirm the line `batch_size = len(pt_offset) - 1` in `loss_canonical_color.py` is a bug in our public code. The correct version should be `batch_size = len(pt_offset)`. The `-1` discards valid samples. Thank you for catching this mistake.

2. **Backbone weight loading for orgfind3d.pth**
The real fine‑tuning loading logic is shown below. We only load weights into `model.backbone` (PointSemSeg backbone), **not the full PointSemSegWithDecoder**:
```python
pretrained_weights = torch.load(args.pretrained_path)
backbone_weights = {
    k.replace('backbone.', ''): v
    for k, v in pretrained_weights["model_state_dict"].items()
    if k.startswith('backbone.') or 'decoder.' not in k
}
model.backbone.load_state_dict(backbone_weights, strict=False)
```
As you observed, keys inside `orgfind3d.pth` are raw PointSemSeg keys without the `backbone.` prefix. The public `train_release.py` does not reproduce this exact loading flow.

3. **Effective batch‑size for contrastive learning**
The official final model was trained with **8 GPUs × 32 = global effective batch size 256**, lr = 5e‑4.

Contrastive learning heavily depends on sufficient negative samples within the batch. Your setup `4 × 24 = 96` is ~2.67× smaller. Smaller batch degrades embedding quality due to insufficient negative samples.
We have received similar reports that reproduction on RTX 3090 cannot match original paper performance.
Two recommended fixes:
- Preferred: Use **gradient accumulation of 3 steps**, effective batch ~256, keep `lr=5e‑4`.
- Alternative (less ideal for contrastive loss): scale learning rate proportionally: `5e‑4 * (96/256) ≈ 1.9e‑4`.

4. **DDP for last‑30‑epoch loss dropping**
You are completely correct. In our real training run we set `find_unused_parameters=True` for DDP, so no error is triggered when canonical‑color loss is dropped in the final 30 epochs. This flag is missing in the current public code.

Again thanks a lot for your careful reproduction and valuable bug reports. We will fix these issues in subsequent commits.

**Comment by JinLi998** [comment]
To summarize your question: we believe batch size is a critical parameter for contrastive learning.
Potential mitigation directions you may refer to include momentum queues such as MoCo / MoCo‑v2 / v3, bank‑based feature libraries, and gradient accumulation. Note that we have not experimented with these alternatives.

Thanks again for reporting the bugs. I will process and upload the 500+ corrupted samples within this week.

**Comment by JinLi998** [comment]
More than 500 corrupted files have been re‑processed and uploaded to:
[https://huggingface.co/PrinterLi/CoSMo3D/resolve/main/forfind3dtrain_cansup.tar.zst?download=true](https://huggingface.co/PrinterLi/CoSMo3D/resolve/main/forfind3dtrain_cansup.tar.zst?download=true)

Decompression command:

```
tar -I zstd -xf forfind3dtrain_cansup.tar.zst
```
